src/tools/dashboard/src/reference.rs

```diff
@@ -0,0 +1,287 @@
+            "--persist-doctests",
+            gen_dir.to_str().unwrap(),
+            "--no-run",
+        ]);
+        cmd.stdout(Stdio::null());
+        cmd.spawn()
+            .expect(&format!("Error: failed to spawn process:\n{:?}", cmd))
+            .wait()
+            .expect(&format!("Error: spawned process did not run:\n{:?}", cmd));
+    }
+}
+
+/// Copies the extracted rust code in `from_dir` to `src/test` following the
+/// hierarchy specified by `rel_map`.
+fn organize_examples(rel_map: &HashMap<PathBuf, PathBuf>, book_dir: &Path, from_dir: &Path) {
+    let to_dir: PathBuf = ["src", "test"].iter().collect();
+    // The names of the extracted examples generated by `rustdoc` have the
+    // format `<path>_<line-num>_<test-num>` where occurrences of '/', '-', and
+    // '.' in <path> are replaced by '_'. This translation is not injective, so
+    // we cannot map those names to the original markdown file path. Instead, we
+    // create a new map that does the same translation.
+    let mut abs_map = HashMap::new();
+    for (path, hierarchy) in rel_map.iter() {
+        abs_map.insert(
+            book_dir.join(path).to_str().unwrap().replace(&['\\', '/', '-', '.'][..], "_"),
+            to_dir.join(hierarchy),
+        );
+    }
+    for dir in from_dir.read_dir().unwrap() {
+        let dir = dir.unwrap().path();
+        // Some directories do not contain tests because the markdown file
+        // instructs `rustdoc` to ignore those tests.
+        if let Some(example) = dir.read_dir().unwrap().next() {
+            let example = example.unwrap().path();
+            copy(&example, &abs_map);
+        }
+    }
+}
+
+/// Copy the file specified by `from` to the corresponding location specified by
+/// `map`.
+fn copy(from: &Path, map: &HashMap<String, PathBuf>) {
+    // First, get the name of the directory where `rustdoc` saved the extracted
+    // code.
+    let file_name = from.parent().unwrap().file_name().unwrap().to_str().unwrap();
+    // Remove `_<line-num>_<test-num>` suffix from the name to lookup the name
+    // in the map.
+    let i = file_name.find("_md").unwrap() + 3;
+    // All extracted examples have different `<path>_<line-num>`, so we can just
+    // ignore the test number.
+    let j = file_name.rfind("_0").unwrap();
+    // The name of the copied file will be `<line-num>.rs`.
+    let to = map[&file_name[..i]].join(format!("{}.rs", &file_name[i + 1..j]));
+    fs::create_dir_all(to.parent().unwrap()).expect(&format!(
+        "Error: failed to create hierarchical directory for extracted test: {:?}",
+        &to
+    ));
+    fs::copy(&from, &to).expect(&format!("Error: failed to copy {:?} to {:?}", &from, &to));
+}
+
+/// Pre-processes the tests in the specified `paths` before running them with
+/// `compiletest`.
+fn preprocess_examples(_paths: Vec<&PathBuf>) {
+    // For now, we will only pre-process the tests that cause infinite loops.
+    // TODO: properly implement this step.
```

Comment:
Reference the issue here.

Reply:
Agreed, this is fine for now, but brittle with hardcoding

Reply:
> Reference the issue here.

Done!

> Agreed, this is fine for now, but brittle with hardcoding

Don't worry. We will come back to it... maybe... 🙃